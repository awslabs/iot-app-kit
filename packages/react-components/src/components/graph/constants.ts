export const GRAPH_COLORS = {
  HEALTH_CRITICAL: '#F45B5B',
  HEALTH_HIGH: '#FA8329',
  HEALTH_LOW: '#8B5CF6',
  HEALTH_MEDIUM: '#FCD62E',
  HEALTH_OFFLINE: '#56575C',
  HEALTH_OK: '#97C2FC',
  HEALTH_UNKNOWN: '#56575C',
  SKY_400: '#38BDF8',
  GRAY_01: '#F8F8F9',
  GRAY_02: '#F4F4F5',
  GRAY_04: '#ECECED',
  GRAY_06: '#E3E3E5',
  GRAY_08: '#DBDBDD',
  GRAY_10: '#D2D3D5',
  GRAY_12: '#CACACD',
  GRAY_14: '#C1C2C5',
  GRAY_16: '#B9BABD',
  GRAY_18: '#B0B1B6',
  GRAY_20: '#A8A9AE',
  GRAY_25: '#93949A',
  GRAY_30: '#7E7F86',
  GRAY_35: '#6A6B71',
  GRAY_40: '#56575C',
  GRAY_45: '#424347',
  GRAY_50: '#2F2F32',
  GRAY_55: '#1B1B1D',
  GRAY_60: '#070708',
  WHITE: '#fff',
  BLUE: '#97C2FC',
  YELLOW: '#FFFF00',
  RED: '#FB7E81',
};

// https://stackoverflow.com/questions/71816702/coordinates-of-dot-on-an-hexagon-path
export const NODE_HEXAGON_HEXAGON_POINTS = [
  0, 1, -0.8660254037844386, 0.5, -0.8660254037844386, -0.5, 0, -1, 0.8660254037844386, -0.5, 0.8660254037844386, 0.5,
];

export const NODE_DIAMOND_DEFAULT_SIZE = 90;
export const NODE_HEXAGON_DEFAULT_SIZE = 100;
export const NODE_ELLIPSE_DEFAULT_SIZE = 80;
export const NODE_RECTANGLE_DEFAULT_SIZE = 70;

export const STYLE_PREFIX = 'iot-app-kit-graph';
