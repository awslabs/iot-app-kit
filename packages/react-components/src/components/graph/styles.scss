@import './lib/css/colors.css';
.ak-graph {
  --background-color: transparent;
  --background-hover-color: var(--color-gray-46);
  --border-radius: 3px;
  --border-color: var(--color-gray-54);
  --border-hover-color: var(--color-gray-54);
  --menu-z-index: 999;
  --icon-color: var(--color-gray-26);
  --icon-hover-color: var(--color-gray-4);
  --icon-active-color: var(--color-gray-4);
  --text-color: var(--color-gray-14);
  --text-hover-color: var(--color-gray-4);
  --text-active-color: var(--color-gray-4);
  --transition: background-color var(--transition-duration) ease-out, border-color var(--transition-duration) ease-out,
    color var(--transition-duration) ease-out, rotate var(--transition-duration) ease-out;
  font-family: var(--font-family);
  font-size: 1px;
  text-rendering: geometricPrecision;

  .ak-graph-content {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    overflow: 'auto';
  }

  .ak-graph-canvas {
    position: absolute;
    top: 0;
    bottom: 0;
    right: 0;
    left: 0;
    z-index: 0;
  }

  .ak-graph-controls {
    position: absolute;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    gap: 1rem;
    width: 1rem;
    height: 100%;
    top: 0;
    right: 5rem;
    z-index: 1;
    pointer-events: none;
  }

  .ak-graph-button {
    box-sizing: content-box;
    display: flex;
    justify-content: center;
    align-items: center;
    width: 2rem;
    height: 2rem;
    background-color: var(--button-background-color);
    border-radius: var(--border-radius);
    box-shadow: 0 0 8px var(--color-gray-50);
    transition: var(--transition);
    pointer-events: all;

    &:hover {
      --button-background-color: var(--button-background-hover-color);
      --button-icon-color: var(--button-icon-hover-color);
    }
  }

  .ak-graph-buttonCenterIcon {
    width: 20rem;
    color: var(--button-icon-color);
    transition: var(--transition);
  }

  .ak-graph-buttonFitIcon {
    width: 18rem;
    color: var(--button-icon-color);
    transition: var(--transition);
  }

  .ak-graph-buttonZoomInIcon {
    width: 16rem;
    color: var(--button-icon-color);
    transition: var(--transition);
  }

  .ak-graph-buttonZoomOutIcon {
    width: 16rem;
    color: var(--button-icon-color);
    transition: var(--transition);
  }

  svg {
    fill: currentcolor;
    shape-rendering: geometricprecision;
  }
}
