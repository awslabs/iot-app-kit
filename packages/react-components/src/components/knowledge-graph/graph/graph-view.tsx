import cytoscape, { Core } from 'cytoscape';
import React, { CSSProperties, forwardRef, MutableRefObject, useRef } from 'react';
import CytoscapeComponent from 'react-cytoscapejs';
import { ErrorBoundary } from 'react-error-boundary';

// This is copied straight from the library, because it's not exported, there's no way to access it, but we are wrapping it.
interface CytoscapeComponentProps {
  id?: string | undefined;
  cy?: ((cy: cytoscape.Core) => void) | undefined;
  style?: CSSProperties | undefined;
  elements: cytoscape.ElementDefinition[];
  layout?: cytoscape.LayoutOptions | undefined;
  stylesheet?: cytoscape.Stylesheet | cytoscape.Stylesheet[] | string | undefined;
  className?: string | undefined;
  zoom?: number | undefined;
  pan?: cytoscape.Position | undefined;
  minZoom?: number | undefined;
  maxZoom?: number | undefined;
  zoomingEnabled?: boolean | undefined;
  userZoomingEnabled?: boolean | undefined;
  boxSelectionEnabled?: boolean | undefined;
  autoungrabify?: boolean | undefined;
  autounselectify?: boolean | undefined;
  panningEnabled?: boolean | undefined;
  userPanningEnabled?: boolean | undefined;
  autolock?: boolean | undefined;
  get?: ((obj: Record<string, any>, key: string) => any) | undefined;
  toJson?: ((obj: Record<string, any>) => any) | undefined;
  diff?: ((objA: Record<string, any>, objB: Record<string, any>) => boolean) | undefined;
  forEach?: (<T>(list: T[], iterator: (value: T, index: number, array: T[]) => void) => void) | undefined;
  headless?: boolean | undefined;
  styleEnabled?: boolean | undefined;
  hideEdgesOnViewport?: boolean | undefined;
  textureOnViewport?: boolean | undefined;
  motionBlur?: boolean | undefined;
  motionBlurOpacity?: number | undefined;
  wheelSensitivity?: number | undefined;
  pixelRatio?: number | string | undefined;
}

const GraphView = forwardRef<Core, CytoscapeComponentProps>(({ ...props }, ref) => {
  const cy = useRef<Core>();
  // This fixes the type issue with the Cytoscape react library that doesn't actually except "refs". So we handle that here,
  // and make it follow react conventions instead of doing it's own thing.
  // istanbul ignore next (this is tested through other tests, but not the negative case)
  const setRef = (core: Core) => {
    (ref as MutableRefObject<Core | null>).current = core;
    cy.current = core;
  };

  return (
    <ErrorBoundary fallback={<div>Error</div>} onError={() => {}}>
      <CytoscapeComponent cy={setRef} layout={{ name: 'cose' }} {...props} />;
    </ErrorBoundary>
  );
});

export default GraphView;
