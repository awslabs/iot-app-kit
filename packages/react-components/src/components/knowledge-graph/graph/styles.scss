@use '@cloudscape-design/design-tokens/index' as awsui;
$prefix: 'iot-app-kit-graph';

.awsui-dark-mode {
  .#{$prefix} {
    /* Background Colors */
    --#{$prefix}-color-background-node: var(--awsui-color-grey-600);
    --#{$prefix}-color-background-node-selected: var(--awsui-color-grey-700);
    --#{$prefix}-color-background-edge: var(--awsui-color-grey-400);
    --#{$prefix}-color-background-edge-selected: var(--awsui-color-grey-700);

    /* Foreground Colors */
    --#{$prefix}-color-node: var(--awsui-color-grey-200);
    --#{$prefix}-color-node-selected: var(--awsui-color-blue-500);
    --#{$prefix}-color-edge: var(--awsui-color-grey-800);
    --#{$prefix}-color-edge-selected: var(--awsui-color-blue-500);

    /* Border Colors */
    --#{$prefix}-color-border-node: var(--awsui-color-grey-400);
    --#{$prefix}-color-border-node-selected: var(--awsui-color-blue-500);
    --#{$prefix}-color-border-edge: var(--awsui-color-grey-400);
    --#{$prefix}-color-border-edge-selected: var(--awsui-color-blue-500);

    /* Line Colors */
    --#{$prefix}-color-edge-line: var(--awsui-color-grey-400);
    --#{$prefix}-color-edge-line-selected: var(--awsui-color-blue-500);

    background-color: var(--awsui-color-grey-800);

    .#{$prefix}-controls {
      background-color: var(--awsui-color-grey-650);
      box-shadow: none;
    }
  }
}

.#{$prefix} {
  /* Background Colors */
  --#{$prefix}-color-background-node: #FFF;
  --#{$prefix}-color-background-node-selected: var(--awsui-color-blue-100);
  --#{$prefix}-color-background-edge: var(--awsui-color-grey-600);
  --#{$prefix}-color-background-edge-selected: var(--awsui-color-blue-100);

  /* Foreground Colors */
  --#{$prefix}-color-node: #000;
  --#{$prefix}-color-node-selected: var(--awsui-color-blue-600);
  --#{$prefix}-color-edge: #FFF;
  --#{$prefix}-color-edge-selected: var(--awsui-color-blue-600);

  /* Border Colors */
  --#{$prefix}-color-border-node: var(--awsui-color-grey-600);
  --#{$prefix}-color-border-node-selected: var(--awsui-color-blue-600);
  --#{$prefix}-color-border-edge: var(--awsui-color-grey-600);
  --#{$prefix}-color-border-edge-selected: var(--awsui-color-blue-600);

  /* Line Colors */
  --#{$prefix}-color-edge-line: var(--awsui-color-grey-600);
  --#{$prefix}-color-edge-line-selected: var(--awsui-color-blue-600);

  font-size: 1px;
  text-rendering: geometricPrecision;
  width: 100%;
  height: 100%;
  min-width: 500px;
  min-height: 500px;
  position: relative;
  background-color: #FFF;

  .#{$prefix}-content {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    overflow: hidden;
  }

  .#{$prefix}-canvas {
    position: absolute;
    width: 100%;
    height: 100%;
    border: solid 2px grey;
    top: 0;
    bottom: 0;
    right: 0;
    left: 0;
    z-index: 0;
  }

  .#{$prefix}-controls {
    position: absolute;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    width: fit-content;
    top: 1rem;
    left: 1rem;
    right: 5rem;
    z-index: 1;
    pointer-events: none;
    padding: 0;
    box-shadow: 2px 5px 8px var(--awsui-color-grey-600);
  }

  .#{$prefix}-control-item {
    border-bottom: 1px solid var(--awsui-color-grey-300);
    padding: 0 4px;
    justify-content: center;
    align-items: center;
  }

  .#{$prefix}-button {
    pointer-events: all;
  }
}
