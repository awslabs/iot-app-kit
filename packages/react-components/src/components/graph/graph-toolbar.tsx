import React, { ReactNode, FC } from 'react';
import { STYLE_PREFIX } from './constants';

export interface ToolbarProps {
  children: ReactNode[];
}

const Toolbar: FC<ToolbarProps> = ({ children }) => {
  return (
    <ul className={`${STYLE_PREFIX}-controls`}>
      {children.map((child, index) => (
        <li key={index} className={`${STYLE_PREFIX}-control-item`}>
          {child}
        </li>
      ))}
    </ul>
  );
};

export default Toolbar;
