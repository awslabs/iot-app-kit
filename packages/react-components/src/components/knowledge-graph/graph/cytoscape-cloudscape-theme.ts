import { STYLE_PREFIX } from './constants';
import { RefObject, useCallback, useEffect, useState } from 'react';

export const colors = {
  nodeBackground: `--${STYLE_PREFIX}-color-background-node`,
  nodeBackgroundSelected: `--${STYLE_PREFIX}-color-background-node-selected`,
  edgeBackground: ` --${STYLE_PREFIX}-color-background-edge`,
  edgeBackgroundSelected: `--${STYLE_PREFIX}-color-background-edge-selected`,

  /* Foreground Colors */
  node: `--${STYLE_PREFIX}-color-node`,
  nodeSelected: `--${STYLE_PREFIX}-color-node-selected`,
  edge: `--${STYLE_PREFIX}-color-edge`,
  edgeSelected: `--${STYLE_PREFIX}-color-edge-selected`,

  /* Border Colors */
  nodeBorder: `--${STYLE_PREFIX}-color-border-node`,
  nodeBorderSelected: `--${STYLE_PREFIX}-color-border-node-selected`,
  edgeBorder: `--${STYLE_PREFIX}-color-border-edge`,
  edgeBorderSelected: `--${STYLE_PREFIX}-color-border-edge-selected`,

  /* Line Colors */
  edgeLine: `--${STYLE_PREFIX}-color-edge-line`,
  edgeLineSelected: `--${STYLE_PREFIX}-color-edge-line-selected`,
};

const useCloudScapeTheme = (container: RefObject<HTMLDivElement>) => {
  /**
   * For simplicty, we'll return a utility function to lookup a color value
   */
  const colorLookup = useCallback(
    (color: keyof typeof colors, fallback?: string) => {
      if (container.current) {
        const c = getComputedStyle(container.current).getPropertyValue(colors[color]);
        // console.log(`Calculated color ${color} to equal ${c}`);
        return c;
      }

      // istanbul ignore next (shouldn't really be a factor)
      return fallback;
    },
    [container.current]
  );

  return colorLookup;
};

const useStylesheet = (container: RefObject<HTMLDivElement>, labelProp = 'label') => {
  const color = useCloudScapeTheme(container);
  const [stylesheet, setStylesheet] = useState<cytoscape.Stylesheet[]>([]);

  useEffect(() => {
    if (color) {
      setStylesheet([
        {
          selector: 'node',
          css: {
            shape: 'round-rectangle',
            label: `data(${labelProp})`,
            width: 'label',
            height: 'label',
            'text-background-padding': '1rem',
            'border-width': '0.25rem',
            'padding-left': '2rem',
            'background-color': color('nodeBackground'),
            color: color('node'),
            'text-background-color': color('node'),
            'border-color': color('nodeBorder'),
            'font-size': '5rem',
            'text-valign': 'center',
          },
        },
        {
          selector: 'node:selected',
          css: {
            'background-color': color('nodeBackgroundSelected'),
            color: color('nodeSelected'),
            'border-color': color('nodeBorderSelected'),
          },
        },
        {
          selector: 'edge',
          css: {
            label: `data(${labelProp})`,
            color: color('edge'),
            width: '0.25rem',
            'text-background-color': color('nodeBorder'),
            'font-size': '5rem',
            'text-background-opacity': 1,
            'text-border-style': 'solid',
            'text-border-opacity': 1,
            'text-border-width': '0.25rem',
            'text-border-color': color('edgeBorder'),
            'text-background-padding': '1rem',
            'line-color': color('edgeLine'),
            'target-arrow-shape': 'chevron',
            'arrow-scale': 0.25,
            'target-arrow-color': color('edgeLine'),
            'curve-style': 'bezier',
          },
        },
        {
          selector: 'edge:selected',
          css: {
            'text-background-color': color('edgeBackgroundSelected'),
            color: color('edgeSelected'),
            'text-border-color': color('edgeBorderSelected'),
            'line-color': color('edgeLineSelected'),
            'target-arrow-color': color('edgeLineSelected'),
          },
        },
      ] as cytoscape.Stylesheet[]);
    }
  }, [color, labelProp]);

  return stylesheet;
};

export default useStylesheet;
