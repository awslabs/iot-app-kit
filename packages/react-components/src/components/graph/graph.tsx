import React, { FC, HTMLAttributes, useCallback, useEffect, useRef } from 'react';
import type { Core, ElementDefinition, LayoutOptions } from 'cytoscape';
import { FitIcon, MinusIcon, PlusIcon, TargetIcon } from './lib/svgs/icons';
import Container from '@awsui/components-react/container';

import CytoscapeComponent from 'react-cytoscapejs';

import './styles.scss';

export interface CytoGraphProps extends HTMLAttributes<HTMLDivElement> {
  graphLayout?: LayoutOptions;
}

export const ZOOM_INTERVAL = 1;

export const Graph: FC<CytoGraphProps> = ({ graphLayout, className, ...props }) => {
  const elements = [
    { data: { id: 'one', label: 'Node 1' }, position: { x: 0, y: 0 } },
    { data: { id: 'two', label: 'Node 2' }, position: { x: 100, y: 0 } },
    { data: { source: 'one', target: 'two', label: 'Edge from Node1 to Node2' } },
  ];

  const cy = useRef<Core>();

  const handleFitClick = useCallback(() => {
    cy.current?.fit();
  }, []);

  const handleCenterClick = useCallback(() => {
    cy.current?.center();
  }, []);

  const handleZoomInClick = useCallback(() => {
    cy.current?.zoom(cy.current.zoom() + ZOOM_INTERVAL);
  }, []);

  const handleZoomOutClick = useCallback(() => {
    cy.current?.zoom(cy.current.zoom() - ZOOM_INTERVAL);
  }, []);

  useEffect(() => {
    cy.current?.resize();
    cy.current?.center();
  }, []);

  // istanbul ignore next
  // Don't care about testing reacts built in ref setting logic.
  const setCy = (c: Core) => {
    cy.current = c;
  };

  return (
    <Container className={`ak-graph ${className || ''}`.trim()} {...props}>
      <CytoscapeComponent
        cy={setCy}
        layout={graphLayout}
        elements={elements as ElementDefinition[]}
        style={{ width: '100vw', height: '100vh' }}
      />
      <div className='ak-graph-controls'>
        <button data-testid='fit-button' className='ak-graph-button' onClick={handleFitClick}>
          <FitIcon className='ak-graph-buttonFitIcon' />
        </button>
        <button data-testid='center-button' className='ak-graph-button' onClick={handleCenterClick}>
          <TargetIcon className='ak-graph-buttonCenterIcon' />
        </button>
        <button data-testid='zoom-in-button' className='ak-graph-button' onClick={handleZoomInClick}>
          <PlusIcon className='ak-graph-buttonZoomInIcon' />
        </button>
        <button data-testid='zoom-out-button' className='ak-graph-button' onClick={handleZoomOutClick}>
          <MinusIcon className='ak-graph-buttonZoomOutIcon' />
        </button>
      </div>
    </Container>
  );
};
