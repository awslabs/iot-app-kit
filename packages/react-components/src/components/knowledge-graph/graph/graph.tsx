import React, { HTMLAttributes, useCallback, useEffect, useRef } from 'react';
import type { Core, ElementDefinition, EventObject } from 'cytoscape';

import useCyEvent from './hooks/useCyEvent';
import GraphView from './graph-view';
import Toolbar from './graph-toolbar';

import './styles.scss';
import { Button } from '@cloudscape-design/components';
import { STYLE_PREFIX } from './constants';
import useStylesheet from './cytoscape-cloudscape-theme';

export interface GraphProps extends HTMLAttributes<HTMLDivElement> {
  elements: ElementDefinition[];
  onNodeSelected?: (e: EventObject) => void;
  onEdgeSelected?: (e: EventObject) => void;
}

export const ZOOM_INTERVAL = 1;

export function Graph({ className, elements, onNodeSelected, onEdgeSelected, ...props }: GraphProps) {
  const cy = useRef<Core>(null);
  const containerRef = useRef<HTMLDivElement>(null);

  const stylesheet = useStylesheet(containerRef);

  const fit = useCallback(() => {
    cy.current?.fit();
  }, []);

  const center = useCallback(() => {
    cy.current?.center();
  }, []);

  const zoomIn = useCallback(() => {
    cy.current?.zoom(cy.current.zoom() + ZOOM_INTERVAL);
  }, []);

  const zoomOut = useCallback(() => {
    cy.current?.zoom(cy.current.zoom() - ZOOM_INTERVAL);
  }, []);

  useEffect(() => {
    cy.current?.resize();
    cy.current?.center();
  }, []);
  console.log('elements: ', elements);

  // TODO: Event modelling needs to be done properly, this is an example
  // istanbul ignore next (No point in testing as this is going to change in next update)
  useCyEvent('tap', 'node', (e) => onNodeSelected && onNodeSelected(e), cy);

  // istanbul ignore next (No point in testing as this is going to change in next update)
  useCyEvent('tap', 'edge', (e) => onEdgeSelected && onEdgeSelected(e), cy);

  return (
    <div ref={containerRef} className={`${STYLE_PREFIX} ${className || ''}`.trim()} {...props}>
      <GraphView ref={cy} stylesheet={stylesheet} elements={elements} style={{ width: '100%', height: '100%' }} />
      <Toolbar>
        <Button
          data-testid='fit-button'
          className={`${STYLE_PREFIX}-button`}
          onClick={fit}
          iconName='zoom-to-fit'
          variant='icon'
        />
        <Button
          data-testid='center-button'
          className={`${STYLE_PREFIX}-button`}
          onClick={center}
          iconName='expand'
          variant='icon'
        />
        <Button
          data-testid='zoom-in-button'
          className={`${STYLE_PREFIX}-button`}
          onClick={zoomIn}
          iconName='zoom-in'
          variant='icon'
        />
        <Button
          data-testid='zoom-out-button'
          className={`${STYLE_PREFIX}-button`}
          onClick={zoomOut}
          iconName='zoom-out'
          variant='icon'
        />
      </Toolbar>
    </div>
  );
}
