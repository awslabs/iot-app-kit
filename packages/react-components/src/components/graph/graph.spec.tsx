import React, { HTMLAttributes, useRef } from 'react';
import { fireEvent, render } from '@testing-library/react';
import { Graph, ZOOM_INTERVAL } from './graph';
import { ElementDefinition } from 'cytoscape';

jest.mock('@awsui/components-react/container', () => (props: HTMLAttributes<HTMLDivElement>) => (
  <div data-mocked='Container' {...props}></div>
));

jest.mock(
  'react-cytoscapejs',
  () =>
    ({
      elements,
      cy: _cy, // We don't want to propagate this ref arg
      ...props
    }: HTMLAttributes<HTMLDivElement> & { elements: ElementDefinition[]; cy: (() => void) | undefined }) =>
      (
        <div data-mocked='CytoscapeComponent' {...props}>
          {JSON.stringify(elements)}
        </div>
      )
);

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useRef: jest.fn(() => ({ current: null })),
}));

jest.mock('./hooks/useCyEvent');

describe('<graph />', () => {
  it('renders default elements', () => {
    const { container } = render(<Graph />);
    expect(container).toMatchSnapshot();
  });

  it('should resize and center on load', () => {
    const useRefMock = useRef as jest.Mock;

    const fakeCy = {
      current: {
        resize: jest.fn(),
        center: jest.fn(),
        fit: jest.fn(),
        zoom: jest.fn(),
      },
    };

    useRefMock.mockReturnValueOnce(fakeCy);

    render(<Graph />);

    expect(fakeCy.current.resize).toHaveBeenCalled();
    expect(fakeCy.current.center).toHaveBeenCalled();
  });

  it('should fit to screen on fit button clicked', async () => {
    const useRefMock = useRef as jest.Mock;

    const fakeCy = {
      current: {
        resize: jest.fn(),
        center: jest.fn(),
        fit: jest.fn(),
        zoom: jest.fn(),
      },
    };

    useRefMock.mockReturnValueOnce(fakeCy);

    const { findByTestId } = render(<Graph />);
    const sut = await findByTestId('fit-button');

    fireEvent.click(sut);

    expect(fakeCy.current.fit).toBeCalled();
  });

  it('should center on screen when center button clicked', async () => {
    const useRefMock = useRef as jest.Mock;

    const fakeCy = {
      current: {
        resize: jest.fn(),
        center: jest.fn(),
        fit: jest.fn(),
        zoom: jest.fn(),
      },
    };

    useRefMock.mockReturnValueOnce(fakeCy);

    const { findByTestId } = render(<Graph />);
    const sut = await findByTestId('center-button');

    fireEvent.click(sut);

    expect(fakeCy.current.center).toBeCalled();
  });

  it('should zoom in when zoom in button clicked', async () => {
    const useRefMock = useRef as jest.Mock;

    const fakeCy = {
      current: {
        resize: jest.fn(),
        center: jest.fn(),
        fit: jest.fn(),
        zoom: jest.fn(() => 1),
      },
    };

    useRefMock.mockReturnValueOnce(fakeCy);

    const { findByTestId } = render(<Graph />);
    const sut = await findByTestId('zoom-in-button');

    fireEvent.click(sut);

    expect(fakeCy.current.zoom).toBeCalledWith(1 + ZOOM_INTERVAL);
  });

  it('should zoom out when zoom out button clicked', async () => {
    const useRefMock = useRef as jest.Mock;

    const fakeCy = {
      current: {
        resize: jest.fn(),
        center: jest.fn(),
        fit: jest.fn(),
        zoom: jest.fn(() => 1),
      },
    };

    useRefMock.mockReturnValueOnce(fakeCy);

    const { findByTestId } = render(<Graph />);
    const sut = await findByTestId('zoom-out-button');

    fireEvent.click(sut);

    expect(fakeCy.current.zoom).toBeCalledWith(1 - ZOOM_INTERVAL);
  });
});
