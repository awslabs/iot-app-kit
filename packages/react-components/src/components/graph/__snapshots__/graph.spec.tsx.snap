// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<graph /> renders default elements 1`] = `
<div>
  <div
    class="ak-graph"
    data-mocked="Container"
  >
    <div
      data-mocked="CytoscapeComponent"
      style="width: 100vw; height: 100vh;"
    >
      [{"data":{"id":"one","label":"Node 1"},"position":{"x":0,"y":0}},{"data":{"id":"two","label":"Node 2"},"position":{"x":100,"y":0}},{"data":{"source":"one","target":"two","label":"Edge from Node1 to Node2"}}]
    </div>
    <div
      class="ak-graph-controls"
    >
      <button
        class="ak-graph-button"
        data-testid="fit-button"
      >
        <svg
          class="ak-graph-buttonFitIcon"
          viewBox="0 0 100 100"
        >
          <path
            d="M30 10H10v20H0V0h30v10ZM70 90h20V70h10v30H70V90Z"
          />
          <path
            d="M80 20v60H20V20h60ZM70 30H30v40h40V30Z"
          />
          <path
            d="M90 30V10H70V0h30v30H90ZM10 70v20h20v10H0V70h10Z"
          />
        </svg>
      </button>
      <button
        class="ak-graph-button"
        data-testid="center-button"
      >
        <svg
          class="ak-graph-buttonCenterIcon"
          viewBox="0 0 100 100"
        >
          <path
            d="M50 31.818c-10.045 0-18.182 8.137-18.182 18.182 0 10.045 8.137 18.182 18.182 18.182 10.045 0 18.182-8.137 18.182-18.182 0-10.045-8.137-18.182-18.182-18.182Zm40.636 13.637c-2.09-18.955-17.136-34-36.09-36.091V0h-9.091v9.364c-18.955 2.09-34 17.136-36.091 36.09H0v9.091h9.364c2.09 18.955 17.136 34 36.09 36.091V100h9.091v-9.364c18.955-2.09 34-17.136 36.091-36.09H100v-9.091h-9.364ZM50 81.818c-17.59 0-31.818-14.227-31.818-31.818 0-17.59 14.227-31.818 31.818-31.818 17.59 0 31.818 14.227 31.818 31.818 0 17.59-14.227 31.818-31.818 31.818Z"
          />
        </svg>
      </button>
      <button
        class="ak-graph-button"
        data-testid="zoom-in-button"
      >
        <svg
          class="ak-graph-buttonZoomInIcon"
          viewBox="0 0 100 100"
        >
          <polygon
            points="55.5555555 0 55.5554688 44.4442708 100 44.4444445 100 55.5555555 55.5554688 55.5554688 55.5555555 100 44.4444445 100 44.4442708 55.5554688 0 55.5555555 0 44.4444445 44.4442708 44.4442708 44.4444445 0"
          />
        </svg>
      </button>
      <button
        class="ak-graph-button"
        data-testid="zoom-out-button"
      >
        <svg
          class="ak-graph-buttonZoomOutIcon"
          viewBox="0 0 100 16"
        >
          <rect
            height="16"
            width="100"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
